agent/taskresource/logrouter/logrouter_unix.go

```diff
@@ -0,0 +1,355 @@
+		err := errors.New(fmt.Sprintf("invalid log router type: %s", logRouter.logRouterType))
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	err := logRouter.createDirectories()
+	if err != nil {
+		err = errors.Wrapf(err, "unable to initialize resource directory %s", logRouter.resourceDir)
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	err = logRouter.generateConfigFile()
+	if err != nil {
+		err = errors.Wrap(err, "unable to generate log router config file")
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// createDirectories creates two directories:
+//  - $(DATA_DIR)/logrouter/$(TASK_ID)/config: used to store log router config file. The config file under this directory
+//    will be mounted to the log router container at an expected path.
+//  - $(DATA_DIR)/logrouter/$(TASK_ID)/socket: used to store the unix socket. This directory will be mounted to
+//    the log router container and it will generate a socket file under this directory. Containers that use log router
+//    will then use this socket to send logs to log router.
+func (logRouter *LogRouterResource) createDirectories() error {
+	configDir := filepath.Join(logRouter.resourceDir, "config")
+	err := logRouter.os.MkdirAll(configDir, os.ModePerm)
+	if err != nil {
+		return errors.Wrap(err, "unable to create config directory")
+	}
+
+	socketDir := filepath.Join(logRouter.resourceDir, "socket")
+	err = logRouter.os.MkdirAll(socketDir, os.ModePerm)
+	if err != nil {
+		return errors.Wrap(err, "unable to create socket directory")
+	}
+	return nil
+}
+
+// generateConfigFile generates a log router config file at $(RESOURCE_DIR)/config/fluent.conf.
+// This contains configs needed by the router container to properly route logs.
+func (logRouter *LogRouterResource) generateConfigFile() error {
+	config, err := logRouter.generateConfig()
+	if err != nil {
+		return errors.Wrap(err, "unable to generate log router config")
+	}
+
+	temp, err := logRouter.ioutil.TempFile(logRouter.resourceDir, tempFile)
+	if err != nil {
+		return err
+	}
+	defer temp.Close()
+	if logRouter.logRouterType == LogRouterTypeFluentd {
+		err = config.WriteFluentdConfig(temp)
+	} else {
+		err = config.WriteFluentBitConfig(temp)
+	}
+	if err != nil {
+		return err
+	}
+
+	err = temp.Chmod(os.FileMode(configFilePerm))
+	if err != nil {
+		return err
+	}
+
+	// Persist the config file to disk.
+	err = temp.Sync()
+	if err != nil {
+		return err
+	}
+
+	confFilePath := filepath.Join(logRouter.resourceDir, "config", "fluent.conf")
+	err = logRouter.os.Rename(temp.Name(), confFilePath)
+	if err != nil {
+		return err
+	}
+
+	seelog.Infof("Generated log router config file at: %s", confFilePath)
+	return nil
+}
+
+// Cleanup performs resource cleanup.
+func (logRouter *LogRouterResource) Cleanup() error {
+	err := logRouter.os.RemoveAll(logRouter.resourceDir)
+	if err != nil {
```

Comment:
is there a way we can force remove them?

Reply:
hmm if we fail to remove the directory, i guess we just can't remove it.. i can't think of why that can happen though